### PR TITLE
Fenrir fixes

### PR DESCRIPTION
**Pull request by embhorn**
* Fix f-3627: test BrokerBufCompare length-fold auth backstop
* Fix f-2359: preserve v5 SUBSCRIBE options through encode/decode
* Fix f-3394: scrub CONNECT credentials from broker rx_buf after auth
* Fix f-6656: scrub v5 CONNACK auth data from client rx_buf

### CODE REVIEW

File: tests/test_broker_connect.c [hunk 1]
```diff
@@ -928,6 +986,90 @@ static BrokerClient* find_broker_client(MqttBroker* broker, const char* id)
...
+    }
+    for (i = 0; i + nlen <= hlen; i++) {
+        if (XMEMCMP(hay + i, needle, (size_t)nlen) == 0) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/* f-3394: after an accepted CONNECT the plaintext credentials must not
```
Comment (aidangarske) [inline tests/test_broker_connect.c]: remove the f-## reference and shorten comment please

File: tests/test_mqtt_client.c [hunk 2]
```diff
@@ -590,6 +590,61 @@ TEST(connect_accepted_connack_latches_v5_props)
     ASSERT_EQ(1024, test_client.packet_sz_max);
 }
 
+/* f-6656: an accepted v5 CONNACK can carry MQTT_PROP_AUTH_DATA - the final
```
Comment (aidangarske) [inline tests/test_mqtt_client.c]: same here

File: tests/test_mqtt_packet.c [hunk 3]
```diff
@@ -1827,6 +1827,39 @@ TEST(encode_subscribe_options_byte_qos2)
     ASSERT_EQ(0x02, tx_buf[rc - 1]);
 }
 
+#ifdef WOLFMQTT_V5
+/* f-2359: the v5 option bits (No Local / Retain As Published / Retain
```
Comment (aidangarske) [inline tests/test_mqtt_packet.c]: same here

File: tests/test_mqtt_packet.c [hunk 4]
```diff
@@ -3463,6 +3496,71 @@ TEST(decode_subscribe_v5_options_reserved_bits_rejected)
     rc = MqttDecode_Subscribe(rx_buf, (int)sizeof(rx_buf), &sub);
     ASSERT_EQ(MQTT_CODE_ERROR_MALFORMED_DATA, rc);
 }
+
+/* f-2359: the v5 SUBSCRIBE options byte carries No Local (bit 2), Retain As
```
Comment (aidangarske) [inline tests/test_mqtt_packet.c]: same here

File: tests/test_broker_connect.c [hunk 5]
```diff
@@ -590,6 +590,64 @@ TEST(connect_v311_binary_password_exact_match_accepted)
     MqttBroker_Free(&broker);
 }
 
+/* Mutation guard for the length-fold backstop in BrokerBufCompare (the
```
Comment (aidangarske) [inline tests/test_broker_connect.c]: shorten this comment please

File: src/mqtt_packet.c [hunk 6]
```diff
@@ -2567,11 +2567,21 @@ int MqttEncode_Subscribe(byte *tx_buf, int tx_buf_len,
 
     /* Encode payload */
     for (i = 0; i < subscribe->topic_count; i++) {
+        byte options;
         topic = &subscribe->topics[i];
         tx_payload += MqttEncode_String(tx_payload, topic->topic_filter);
+        options = (byte)topic->qos;
```
Comment (aidangarske) [inline src/mqtt_packet.c:2573]: **🟠 [Medium] v5 SUBSCRIBE encoder folds uninitialized/unvalidated MqttTopic.sub_options into the options byte, emitting reserved or…** · _Logic / State Management (CWE-20 input validation, CWE-457…_

All three scan modes independently flagged the same issue at the v5 SUBSCRIBE encoder, so confidence is raised to High (all modes agreed on Medium severity — no severity conflict). BEFORE this PR the SUBSCRIBE payload options byte was written only from topic->qos, so bits 2-7 were always 0 regardless of how the caller initialized MqttTopic. AFTER the PR, for a v5 SUBSCRIBE the loop ORs in `topic->sub_options & MQTT_SUBSCRIBE_OPTIONS_MASK` (mask 0x3C, bits 2-5) with no validation. MqttClient_Subscribe unconditionally sets subscribe->protocol_level = client->protocol_level (src/mqtt_client.c:2491), and in a WOLFMQTT_V5 build MqttClient_Init defaults the client to v5 (MQTT_CONNECT_PROTOCOL_LEVEL == MQTT_CONNECT_PROTOCOL_LEVEL_5), so the fold is active for every v5 client. Two failure modes: (1) sub_options is a brand-new struct member, so existing v5 callers that populated only topic_filter and qos on a non-zero-initialized MqttTopic (previously sufficient) now emit uninitialized bits 2-5 (CWE-457) — a silent behavioral/contract change on upgrade; (2) an intentional or garbage value can set Retain Handling = 3 (0x30), which MQTT v5 §3.8.3.1 reserves and which this same file's decoder (MqttDecode_Subscribe) rejects, so the encoder can emit a Malformed Packet its own decoder would reject and a spec-strict broker MUST disconnect; other combinations silently alter subscription semantics (No Local / Retain As Published). The PR mitigates the two bundled stack-local examples (mqttsimple.c, websocket_client.c add XMEMSET(topics, 0, ...)) and documents the zero-init requirement in…

**Fix:** Validate topic->sub_options before encoding v5 SUBSCRIBE options: reject bits outside MQTT_SUBSCRIBE_OPTIONS_MASK and reject `((topic->sub_options >> 4) & 0x03) == 0x03` (Retain Handling = 3) with MQTT_CODE_ERROR_BAD_ARG / MQTT_CODE_ERROR_MALFORMED_DATA so an uninitialized or invalid field cannot produce an on-wire Malformed Packet and the encoder stays symmetric with the decoder. Consider an explicit opt-in flag on MqttSubscribe (e.g. use_sub_options) so legacy v5 callers that only set topic_filter/qos keep emitting QoS-only options. Add an encode-side regression test for Retain Handling = 3, zero the remaining in-tree stack MqttTopic caller in IDE/ARDUINO/wolfmqtt_client/wolfmqtt_client.ino, and call out the new MqttTopic zero-init requirement in the ChangeLog/migration notes.

File: src/mqtt_packet.c [hunk 7]
```diff
@@ -2567,11 +2567,21 @@ int MqttEncode_Subscribe(byte *tx_buf, int tx_buf_len,
...
+    #ifdef WOLFMQTT_V5
+        /* [MQTT v5 3.8.3.1] merge No Local, Retain As Published and Retain
```
Comment (aidangarske) [inline src/mqtt_packet.c:2575]: **🔵 [Low] Broker decodes v5 sub_options but never honors No Local / Retain As Published / Retain Handling** · _API / Completeness_

The decode side now populates `topic->sub_options` (bits 2-5) so the options byte round-trips, but src/mqtt_broker.c never reads sub_options (no references to sub_options / no_local / retain-handling anywhere in the broker). So the broker preserves these bits into the caller's topic buffer yet does not enforce No Local loop-prevention, Retain As Published, or Retain Handling on delivery. This is not a regression (the broker never honored them before) and matches the stated scope of preserving v5 SUBSCRIBE options through encode/decode, but it is easy for a future reader to assume decode implies enforcement. Verified: MqttDecode_Subscribe assigns topic->sub_options at src/mqtt_packet.c:2580, and a grep for sub_options / no_local / retain_handling in src/mqtt_broker.c returns no matches.

**Fix:** No code change required for this PR; confirm decode-only preservation is the intended scope and add a short ChangeLog note or one-line comment that the broker currently stores but does not act on these v5 subscription options. Track broker-side enforcement of No Local / Retain Handling as separate follow-up work.

**Review by aidangarske — COMMENTED** [review]
## Skoll Multi-Scan Review**Overall recommendation:** `COMMENT`
**Findings:** 3 total — 3 posted, 0 skipped
**2 finding(s) posted as inline comments** (see file-level comments below)
**1 finding(s) not tied to a diff line** (full detail below)

### Posted findings
- [Medium] [review+review-security+bugs] **v5 SUBSCRIBE encoder folds uninitialized/unvalidated MqttTopic.sub_options into the options byte, emitting reserved or garbage option bits on the wire** — `src/mqtt_packet.c:2417-2436`
- [Low] [review] **Broker decodes v5 sub_options but never honors No Local / Retain As Published / Retain Handling** — `src/mqtt_packet.c:2575-2581`

### Findings not tied to a diff line

#### Credential/CONNACK scrub changes reviewed clean (informational)
**File:** `src/mqtt_broker.c:4939-4957, src/mqtt_client.c:1970-1990`
**Function:** `BrokerHandle_Connect / MqttClient_Connect`
**Severity:** Info
**Category:** Zeroization

The two new zeroization sites were audited for common failure modes and independently verified against the code as correct. Broker: BROKER_FORCE_ZERO(bc->rx_buf, rx_len) is in-bounds because rx_len is the MqttPacket_Read return, bounded by the rx buffer size; the dynamic-mode password scrub uses password_len + 1, matching the WOLFMQTT_MALLOC(src_len + 1) allocation and NULL-guarded; static-mode scrubs the full BROKER_MAX_PASSWORD_LEN array; password_len = 0 and the password scrub are correctly gated on WOLFMQTT_BROKER_AUTH. No double-free/UAF: the pointer is unchanged so the later BrokerClient_Free scrub+free path still works. Client: the whole-rx_buf scrub is safe because v5 CONNACK properties are consumed synchronously (MqttClient_WaitType delivers and frees the property list before the scrub) and no live pointer into rx_buf remains; ack.return_code is a scalar and survives; the multithread path takes lockRecv non-nested (no deadlock) and mirrors the existing MqttClient_Auth hardening. No defect found; recorded for traceability.

**Recommendation:** No action required.

_Referenced code: `src/mqtt_broker.c:4939-4957, src/mqtt_client.c:1970-1980` (11 lines)_

---

*Review generated by [Skoll](https://github.com/wolfSSL/fenrir/tree/main/skoll)*

**Review by aidangarske — APPROVED** [review]
